src/instructlab/process/process.py

```diff
@@ -264,51 +317,46 @@ def all_processes_running(pids: list[int]) -> bool:
     return all(psutil.pid_exists(pid) for pid in pids)
 
 
-def stop_process(local_uuid, remove=True):
+def stop_process(local_uuid: str, remove=True):
     """
     Stop a running process.
 
     Args:
         local_uuid (str): uuid of the process to stop.
     """
-    process_registry = load_registry()
+    process_registry = ProcessRegistry().load()
     # we should kill the parent process, and also children processes.
-    pid = process_registry.processes[local_uuid]["pid"]
-    children_pids = process_registry.processes[local_uuid]["children_pids"]
-    all_processes = [pid] + children_pids
-    for process in all_processes:
+    if local_uuid not in process_registry.processes:
+        logger.warning("Process not found.")
+        return
+
+    process = process_registry.processes[local_uuid]
+    for pid in process.pids:
         try:
-            os.kill(process, signal.SIGKILL)
-            logger.info(f"Process {process} terminated.")
+            os.kill(pid, signal.SIGKILL)
+            logger.info(f"Process {pid} terminated.")
         except (ProcessLookupError, PermissionError):
-            logger.warning(
-                f"Process {process} was not running or could not be stopped."
-            )
+            logger.warning(f"Process {pid} was not running or could not be stopped.")
+
     if remove:
-        process_registry.processes.pop(local_uuid, None)
+        process_registry.remove(local_uuid)
     else:
         # since we just killed the processes, we cannot depend on it to update itself, mark as done and set end time
-        process_registry.processes[local_uuid]["status"] = ILAB_PROCESS_STATUS.DONE
-        process_registry.processes[local_uuid]["end_time"] = datetime.now().isoformat()
-    save_registry(process_registry=process_registry)
+        process.complete(ILAB_PROCESS_STATUS.DONE)
+    process_registry.persist()
 
 
-def update_status(local_uuid, status):
+def complete_process(local_uuid: str, status):
     """
     Updates the status of a process.
 
     Args:
         local_uuid (str): uuid of the process to stop.
     """
 
-    process_registry = load_registry()
-    entry = process_registry.processes.get(str(local_uuid), {})
-    if entry:
-        process_registry.processes[str(local_uuid)]["end_time"] = (
-            datetime.now().isoformat()
-        )
-        process_registry.processes[str(local_uuid)]["status"] = status
-        save_registry(process_registry=process_registry)
+    process = ProcessRegistry().load().processes.get(local_uuid, None)
+    if process:
+        process.complete(status)
```

Comment:
ah there's a bug here, forgot to persist!

Reply:
(We should probably have a transaction semantics managed by the registry to avoid this kind of mistakes in the future.)